cocotb/share/lib/simulator/simulatormodule.cpp

```diff
@@ -122,14 +122,6 @@ namespace {
         type.tp_richcompare = gpi_hdl_richcompare<gpi_hdl>;
         return type;
     }
-
-    // these will be initialized later, once the members are all defined
-    template<>
-    PyTypeObject gpi_hdl_Object<gpi_sim_hdl>::py_type;
-    template<>
-    PyTypeObject gpi_hdl_Object<gpi_iterator_hdl>::py_type;
-    template<>
-    PyTypeObject gpi_hdl_Object<gpi_cb_hdl>::py_type;
```

Comment:
I remember needing to forward-declare these in order to reference them below. Perhaps the references were removed in a later commit.

Does msvc reject these?

Reply:
```
  cocotb\share\lib\simulator\simulatormodule.cpp(964): error C2086: 'PyTypeObject `anonymous namespace'::gpi_hdl_Object<GpiObjHdl *>::py_type': redefinition
  cocotb\share\lib\simulator\simulatormodule.cpp(128): note: see declaration of 'py_type'
```
Without forward-decleration MSVC is happy but clang complains:
```
  cocotb/share/lib/simulator/simulatormodule.cpp:956:43: error: explicit specialization of 'py_type' after instantiation
  PyTypeObject gpi_hdl_Object<gpi_sim_hdl>::py_type = []() -> PyTypeObject {
                                            ^
  cocotb/share/lib/simulator/simulatormodule.cpp:485:59: note: implicit instantiation first required here
      if (Py_TYPE(pSigHdl) != &gpi_hdl_Object<gpi_sim_hdl>::py_type) {
```
GCC is fine either way

Reply:
Any idea which compiler is following the standard here?

Reply:
I think the relevant part from spec is §14.7.3 [temp.expl.spec]/p6 (emphasis added):
> If a template, a member template or a member of a class template is explicitly specialized then **that specialization shall be declared before the first use of that specialization that would cause an implicit instantiation to take place, in every translation unit in which such a use occurs**; no diagnostic is required.

I updated the code to move the template before its first use and now it works fine on all compilers, see [`078dc8b`](https://github.com/cocotb/cocotb/pull/1798/commits/078dc8b3cccae2fce9e8feaa1d862443a500c6fe).

Reply:
I asked [on StackOverflow here](https://stackoverflow.com/q/61770519/102441). To me this looks like an MSVC bug, the existing code already contains the declaration.

Reply:
@eric-wieser, I read your comment on community [bug report](https://developercommunity.visualstudio.com/content/problem/89837/vc2017-a-declaration-of-variable-template-is-alway.html) (which was mysteriously closed by some sort of a canned response..). However, [similar report](https://developercommunity.visualstudio.com/content/problem/319447/explicit-specialization-of-static-data-member-inco.html) was made where it was explained that it has been fixed in VS2019 16.5; but since it is a breaking change, it will require `/permissive-` to compile in standard conformance mode.

library-free repro:

```c++
// specialization-declaration.cpp
template<typename X> class A { };

template<typename Y> class B
{
public:
    static A<Y> j;
};

template<> A<int> B<int>::j;
A<int> y = B<int>::j;
template<> A<int> B<int>::j = { };
```
with:
```cmd
cl /c .\specialization-declaration.cpp
```
gives `C2086` errors, while in std conformance mode:

```cmd
cl /c /permissive- .\specialization-declaration.cpp
```
compiles fine:
```
Microsoft (R) C/C++ Optimizing Compiler Version 19.26.28806 for x86
Copyright (C) Microsoft Corporation.  All rights reserved.

specialization-declaration.cpp
```
(Visual Studio 2019, 16.6.3)

Reply:
@am11 Thank you, adding `/permissive-` fixed the compile error.